File: solutions/0506. Relative Ranks/0506.cpp
// Time complexity: O(n+m)
// Space complexity: O(n+m)
class Solution {
 public:
  vector<string> findRelativeRanks(const vector<int>& score) {
    int n = score.size();
    int m = *max_element(score.begin(), score.end());

    vector<int> scoreToIndex(m + 1, 0);
    for (int i = 0; i < n; ++i)
      scoreToIndex[score[i]] = i + 1;

    const vector<string> MEDALS = {"Gold Medal", "Silver Medal", "Bronze Medal"};

    vector<string> ans(n);
    int place = 1;
    for (int i = m; i >= 0; --i)
      if (scoreToIndex[i] != 0) {
        int originalIndex = scoreToIndex[i] - 1;
        if (place < 4)
          ans[originalIndex] = MEDALS[place - 1];
        else
          ans[originalIndex] = to_string(place);
        ++place;
      }

    return ans;   
  }
};

File: solutions/0506. Relative Ranks/0506.py
# Time complexity: O(n+m)
# Space complexity: O(n+m)
class Solution:
  def findRelativeRanks(self, score: List[int]) -> List[str]:
    n = len(score)
    m = max(score)

    scoreToIndex = [0] * (m + 1)
    for i in range(n):
      scoreToIndex[score[i]] = i + 1

    MEDALS = ["Gold Medal", "Silver Medal", "Bronze Medal"]

    ans = [""] * n
    place = 1
    for i in range(m, -1, -1):
      if scoreToIndex[i] != 0:
        originalIndex = scoreToIndex[i] - 1
        if place < 4:
          ans[originalIndex] = MEDALS[place - 1]
        else:
          ans[originalIndex] = str(place)

    return ans